lib/beacon/broadcast.go

```diff
@@ -150,77 +150,74 @@ func (w *broadcastWriter) Serve() {
 		}
 
 		if success > 0 {
-			w.setError(nil)
+			w.SetError(nil)
 		}
 	}
-}
 
-func (w *broadcastWriter) Stop() {
-	w.connMut.Lock()
-	if w.conn != nil {
-		w.conn.Close()
-	}
-	w.connMut.Unlock()
+	return nil
 }
 
 func (w *broadcastWriter) String() string {
 	return fmt.Sprintf("broadcastWriter@%p", w)
 }
 
 type broadcastReader struct {
-	port    int
-	outbox  chan recv
-	conn    *net.UDPConn
-	connMut sync.Mutex
-	errorHolder
+	util.ServiceWithError
+	port   int
+	outbox chan recv
 }
 
-func (r *broadcastReader) Serve() {
+func (r *broadcastReader) serve(stop chan struct{}) error {
 	l.Debugln(r, "starting")
 	defer l.Debugln(r, "stopping")
 
 	conn, err := net.ListenUDP("udp4", &net.UDPAddr{Port: r.port})
 	if err != nil {
 		l.Debugln(err)
-		r.setError(err)
-		return
+		return err
 	}
 	defer conn.Close()
 
-	r.connMut.Lock()
-	r.conn = conn
-	r.connMut.Unlock()
-
+	readResChan := make(chan readRes)
 	bs := make([]byte, 65536)
-	for {
-		n, addr, err := conn.ReadFrom(bs)
-		if err != nil {
-			l.Debugln(err)
-			r.setError(err)
-			return
-		}
-
-		r.setError(nil)
 
-		l.Debugf("recv %d bytes from %s", n, addr)
+	for {
+		go func() {
+			for {
+				n, addr, err := conn.ReadFrom(bs)
```

Comment:
This is a race against the copy out of bs in the loop below.

I think it would be cleaner to not have this goroutine and break the loop by closing the socket as previously? If having a separate stop method doesn't fit in the new architecture, this could maybe be as simple as launching a `go func() { <-stop; conn.Close() }` at the beginning?

Reply:
Done